### PR TITLE
Make statusToAdmin in *Action classes a List #7112

### LINKED ISSUE #7112: Make statusToAdmin in *Action classes a List
Right now, statusToAdmin is a String, which is annoying whenever multiple statuses needs to be sent, more so if it happens inside separate ifs. 

Changing it into a List would make this much more convenient. However, this is a large-scale refactoring, that will touch all *Action classes.

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #7112

### CODE REVIEW

**Comment by whipermr5** [comment]
@sujeet14108 Sorry for the delay 😅 In the meantime can you resolve the merge conflict?

**Comment by sujeet14108** [comment]
@whipermr5  Have a look :)

File: src/main/java/teammates/common/util/LogMessageGenerator.java [hunk 1]
```diff
@@ -43,7 +44,9 @@ public String generatePageActionLogMessage(String url, Map<String, String[]> par
             updateInfoForNormalUser(builder, currUser, userAccount);
         }
 
-        builder.withLogMessage(logMessage);
+        String message = statusToAdmin.isEmpty() ? null : StringHelper.join(
+                "", statusToAdmin.toArray(new String[statusToAdmin.size()]));
+        builder.withLogMessage(message);
```
Comment (whipermr5) [inline src/main/java/teammates/common/util/LogMessageGenerator.java]: Please edit the `Builder` instead.

File: src/main/java/teammates/ui/controller/Action.java [hunk 2]
```diff
@@ -701,4 +701,14 @@ private boolean isMasqueradeModeRequested(AccountAttributes loggedInUser, String
     protected void excludeStudentDetailsFromResponseParams() {
         regkey = null;
     }
+
+    protected void appendToStatus(String message) {
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/Action.java]: I don't think this method is necessary. Instead, modify all the places that use it such that they only call `statusToAdmin.add` once, after the entire message has been prepared.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionCopyActionTest.java [hunk 3]
```diff
@@ -132,7 +133,7 @@ public void testExecuteAndPostProcess() {
                              + "instructorFeedbackQuestionCopy|||true|||"
                              + "Instructor|||Instructor 1 of Course 1|||"
                              + "idOfInstructor1OfCourse1|||instr1@course1.tmt|||"
-                             + "|||/page/instructorFeedbackQuestionCopy";
+                             + "Unknown|||/page/instructorFeedbackQuestionCopy";
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionCopyActionTest.java]: Why was this changed?

**Comment by whipermr5** [comment]
@sujeet14108 Let's not forget this PR :)

**Comment by whipermr5** [comment]
@sujeet14108 Let me know when this is ready for review :)

**Comment by sujeet14108** [comment]
@whipermr5 ready for review :)

**Comment by sujeet14108** [comment]
@whipermr5 Please review :)

**Comment by sujeet14108** [comment]
@whipermr5 Please review :)

File: src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java [hunk 4]
```diff
@@ -19,6 +19,7 @@
 
 public class AdminEmailComposeSendAction extends Action {
 
+    private String status;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java]: `statusMessageToAdmin` would be a better name. Also update the other instances.

File: src/main/java/teammates/common/util/LogMessageGenerator.java [hunk 5]
```diff
@@ -28,11 +29,11 @@
      * @param currUser login information generated by {@link teammates.logic.api.GateKeeper}
      * @param userAccount authentication user account generated by action
      * @param unregisteredStudent authentication unregisteredStudent attributes generated by action
-     * @param logMessage log message to show to admin
+     * @param logMessages log message to show to admin
```
Comment (whipermr5) [inline src/main/java/teammates/common/util/LogMessageGenerator.java]: The description hasn't been updated yet

File: src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java (hunk 4 above)
Comment (sujeet14108) [reply src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java]: Done

File: src/main/java/teammates/common/util/LogMessageGenerator.java (hunk 5 above)
Comment (sujeet14108) [reply src/main/java/teammates/common/util/LogMessageGenerator.java]: `list of log message to show to admin` would this suffice ?

Comment (whipermr5) [reply src/main/java/teammates/common/util/LogMessageGenerator.java]: Yes that's good

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java [hunk 6]
```diff
@@ -62,15 +63,16 @@ protected ActionResult execute() throws EntityDoesNotExistException {
...
         excludeStudentDetailsFromResponseParams();
 
-        if (statusToAdmin == null || statusToAdmin.trim().isEmpty()) {
-            statusToAdmin = studentInfo;
+        if (statusToAdmin == null || statusToAdmin.isEmpty()) {
+            statusMessageToAdmin += studentInfo;
         } else {
-            statusToAdmin += "<br><br>" + studentInfo;
+            statusMessageToAdmin += "<br><br>" + studentInfo;
         }
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: The original behaviour is to append to the existing status message. This approach adds a new status message to the list, which is a change in behaviour.

File: src/main/java/teammates/ui/controller/StudentProfilePictureUploadAction.java [hunk 7]
```diff
@@ -124,15 +127,15 @@ private void deletePicture(BlobKey blobKey) {
         try {
             logic.deletePicture(blobKey);
         } catch (BlobstoreFailureException bfe) {
-            statusToAdmin = Const.ACTION_RESULT_FAILURE
+            statusMessageToAdmin += Const.ACTION_RESULT_FAILURE
                           + " : Unable to delete profile picture (possible unused picture with key: "
                           + blobKey.getKeyString() + " || Error Message: "
                           + bfe.getMessage() + Const.EOL;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/StudentProfilePictureUploadAction.java]: 8-space indentation.

File: src/main/java/teammates/ui/controller/StudentProfilePictureUploadAction.java [hunk 8]
```diff
@@ -23,6 +23,8 @@
  *         that was just uploaded.
  */
 public class StudentProfilePictureUploadAction extends Action {
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/StudentProfilePictureUploadAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java [hunk 9]
```diff
@@ -19,6 +19,7 @@
 
 public class AdminEmailComposeSendAction extends Action {
 
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/InstructorFeedbackRemindParticularStudentsAction.java [hunk 10]
```diff
@@ -6,6 +6,7 @@
 import teammates.common.util.StatusMessageColor;
 
 public class InstructorFeedbackRemindParticularStudentsAction extends Action {
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackRemindParticularStudentsAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java [hunk 11]
```diff
@@ -22,6 +22,7 @@
  * Action: Create a new {@link FeedbackResponseCommentAttributes}.
  */
 public class InstructorFeedbackResponseCommentAddAction extends Action {
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentDeleteAction.java [hunk 12]
```diff
@@ -12,6 +12,7 @@
  * Action: Delete {@link FeedbackResponseCommentAttributes}.
  */
 public class InstructorFeedbackResponseCommentDeleteAction extends InstructorFeedbackResponseCommentAbstractAction {
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentDeleteAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentEditAction.java [hunk 13]
```diff
@@ -20,6 +20,8 @@
  * Action: Edit {@link FeedbackResponseCommentAttributes}.
  */
 public class InstructorFeedbackResponseCommentEditAction extends InstructorFeedbackResponseCommentAbstractAction {
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentEditAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 14]
```diff
@@ -18,17 +18,18 @@
  * his registration key with another student's google account.
  */
 public class StudentCourseJoinAction extends Action {
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Where is this initialised?

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java [hunk 15]
```diff
@@ -26,6 +26,7 @@
 public class StudentCourseJoinAuthenticatedAction extends CourseJoinAuthenticatedAbstractAction {
 
     private static final Logger log = Logger.getLogger();
+    private String statusMessageToAdmin;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: Where is this initialised?

File: src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java [hunk 16]
```diff
@@ -75,8 +77,11 @@ public void generateLogMessage_normalPageAction() {
         String logMessage = "TEAMMATESLOG|||instructorHomePage|||instructorHomePage|||true|||Unknown|||Unknown"
                             + "|||Unknown|||Unknown|||Not authorized|||/page/instructorHomePage";
 
+        List<String> statusToUser = new ArrayList<String>();
+        statusToUser.add("Not authorized");
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java]: `List<String> statusToUser = Arrays.asList("Not authorized");`

File: src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java [hunk 17]
```diff
@@ -104,9 +109,11 @@ public void generateLogMessage_normalPageAction() {
                 .withGoogleId("unknownGoogleId")
                 .build();
 
+        statusToUser.removeAll(statusToUser);
+        statusToUser.add("Join Course");
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java]: `statusToUser = Arrays.asList("Join Course");`

File: src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java [hunk 18]
```diff
@@ -120,9 +127,11 @@ public void generateLogMessage_normalPageAction() {
                      + "|||googleId|||Unknown|||Try student home|||" + url;
         UserType userType = new UserType("googleId");
 
+        statusToUser.removeAll(statusToUser);
+        statusToUser.add("Try student home");
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java]: Same here.

File: src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java [hunk 19]
```diff
@@ -135,8 +144,10 @@ public void generateLogMessage_normalPageAction() {
         userType.isStudent = true;
         AccountAttributes acc = new AccountAttributes("googleId", "david", false, "david@email.com", "NUS");
 
+        statusToUser.removeAll(statusToUser);
+        statusToUser.add("View Result");
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java]: Same here.

File: src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java [hunk 20]
```diff
@@ -189,9 +200,11 @@ public void generateLogMessage_normalPageAction() {
         logMessage = "TEAMMATESLOG|||instructorCoursesPage|||instructorCoursesPage|||true|||Instructor(M)|||david"
                      + "|||anotherGoogleId|||david@email.com|||View comments|||/page/instructorCoursesPage";
 
+        statusToUser.removeAll(statusToUser);
+        statusToUser.add("View comments");
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/LogMessageGeneratorTest.java]: Same here.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java (hunk 6 above)
Comment (sujeet14108) [reply src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: This our new approach is calling statusToAdmin.add once, after the entire message has been prepared. 
Is it change in behaviour ?

Comment (whipermr5) [reply src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: That's assuming `statusToAdmin` is only modified here. Consider the case where `statusToAdmin` is neither null nor empty upon reaching this part of the code. The new behaviour would actually result in two messages, when the original intention was for one message.

Comment (sujeet14108) [reply src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: We are not appending to `statusToAdmin` anywhere. Making ` statusMessageToAdmin` the entire message and then put it in statusToAdmin

Comment (sujeet14108) [reply src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: What's the solution to this ? :sweat_smile:

Comment (sujeet14108) [reply src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: @whipermr5 Any thought on this?

**Comment by wkurniawan07** [comment]
Closing due to an inactivity for a very long period. If anyone is interested to continue the work done here, feel free to reopen/request for reopen.
